rosetta-source/src/main/rosetta/event-common-func.rosetta

```diff
@@ -1661,6 +1661,173 @@ func Qualify_Shaping: <"The qualification of a shaping event from the fact that
+			empty
+			
+	alias newPriceQuantity: <"Create the new price quantity object.">
+		Create_PriceQuantity( [ newPrice ], changeQuantity, empty )
+		
+	alias haircutPercentage:
+        newCashValue / (newAssetQuantity * newAllinPrice)
+	
+	set instruction -> split -> breakdown:
+		[ Create_TerminationInstruction( tradeState ) ]
+	
+	add instruction -> split -> breakdown: <"Sets the second part of the split to be a new contract with a new asset payout.">
+		[
+		Create_PrimitiveInstruction(
+			empty, empty, empty, empty,
+			Create_QuantityChangeInstruction(
+				[newPriceQuantity],
+				QuantityChangeDirectionEnum -> Replace,
+				empty
+				),
+			empty, empty, 
+			Create_RepriceTermsChangeInstruction(
+				tradeState -> trade -> tradableProduct -> product -> contractualProduct,
+				haircutPercentage,
+				effectiveRepriceDate
+				),
+			empty, empty, empty, empty
+			)
+		]
+		
+		
+func Create_RepriceTermsChangeInstruction: <"Creates the relevant terms change primitive instruction object for repricing a contractual product, which consists in the same terms as the original contractual product 
+											but with a new collateral price that may also result in modifing the cash or collateral quantity.">
+	inputs:
+		contractualProduct ContractualProduct (1..1) <"The original contractual product to be repriced.">
+		newHaircutPercentage number (1..1) <"The new haricut value based on cash adjustment.">
+		effectiveRepriceDate AdjustableOrRelativeDate (1..1) <"The date to close and open a new position.">
+	output:
+		termsChangeInstruction TermsChangeInstruction (1..1) <"The relevant primitive instruction for the repricing.">
+
+	set termsChangeInstruction -> product -> contractualProduct:
+		contractualProduct
+	set termsChangeInstruction -> product -> contractualProduct -> economicTerms -> collateral -> collateralProvisions -> eligibleCollateral -> criteria -> treatment ->
+		valuationTreatment -> haircutPercentage:
+		newHaircutPercentage
+	set termsChangeInstruction -> product -> contractualProduct -> economicTerms -> effectiveDate:
+		effectiveRepriceDate
+		
+func Qualify_Reprice:
+	[qualification BusinessEvent]
+	inputs:
+		businessEvent BusinessEvent (1..1)
+	output: is_event boolean (1..1)
+	alias beforeTradePrice: businessEvent -> instruction only-element -> before -> trade -> tradableProduct -> tradeLot->priceQuantity
+
+	alias openTradePrice: FilterOpenTradeStates( businessEvent -> after ) only-element -> trade -> tradableProduct -> tradeLot->priceQuantity
+
+	alias closedTradePrice: FilterClosedTradeStates( businessEvent -> after )
+	
+	alias beforeEconomicterms: businessEvent -> instruction only-element -> before -> trade -> tradableProduct -> product -> contractualProduct -> economicTerms
+	alias openEconomicTerms: FilterOpenTradeStates( businessEvent -> after ) only-element -> trade -> tradableProduct -> product -> contractualProduct -> economicTerms
+	alias closedTradeState: FilterClosedTradeStates( businessEvent -> after )
```

Comment:
Is the tradeState -> state set to closedState then for the after trade? I don't think this is correct?